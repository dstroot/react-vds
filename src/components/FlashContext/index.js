import React, { useReducer, useContext, useMemo } from 'react';

/**
 * Context lets us pass values deep into the component tree. This
 * example allows us to create a "flash message" that can be initiated
 * and viewed from anywhere in our app. First lets create a new context:
 */

const FlashContext = React.createContext({});

/**
 * Using a functional or class component we could pass (via context)
 * both the state and functions to update state. Howevere, using
 * a reducer it's cleaner because state is managed through actions,
 * and we can pass state and the dispatch function via contaxt so you
 * can read the state and dispatch updates from anywhere in you app.
 * https://medium.freecodecamp.org/hooked-on-hooks-how-to-use-reacts-usereducer-2fe8f486b963
 * https://kentcdodds.com/blog/how-to-optimize-your-context-value
 * https://www.robinwieruch.de/react-state-usereducer-usestate-usecontext/
 * https://hswolff.com/blog/how-to-usecontext-with-usereducer/
 */

const initialState = {
  showFlash: false, // Show flash message or not
  message: '', // The message to display
  style: '', // Notification type: "success", "error" or "warning"
};

const reducer = (state, action) => {
  switch (action.type) {
    case 'flash':
      return {
        ...state,
        showFlash: true,
        message: action.message,
        style: action.style,
      };
    case 'dismiss':
      return {
        ...state,
        showFlash: false,
        message: '',
        style: '',
      };
    default:
      throw new Error('Unexpected action');
  }
};

/**
 * Now lets create a hook that simply returns the context value.
 * We can destructure the context array into state and the
 * dispatch function like so:
 *
 * const [state, dispatch] = useFlashContext();
 *
 * This makes it simple to use this context anywhere in your app.
 */

export const useFlashContext = () => {
  const contextValue = useContext(FlashContext);
  if (!contextValue) {
    throw new Error('useFlashContext must be used within a FlashProvider');
  }
  return contextValue;
};

/**
 * Finally let's export a context provider so that we can
 * wrap our application with this provider. This is what actually
 * give us access to the context everywhere in our app.
 */

export const FlashProvider = ({ children }) => {
  // useReducer gives us an array - the current state and
  // the dispatch function [state, dispatch] which we share
  // via context. Passing output of useReducer to context
  // is the "magic".
  const contextValue = useReducer(reducer, initialState);

  // memo so we don't re-render every time app renders
  const value = useMemo(() => {
    return contextValue;
  }, [contextValue]);

  return (
    <FlashContext.Provider value={value}>{children}</FlashContext.Provider>
  );
};
