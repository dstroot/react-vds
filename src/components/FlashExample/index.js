import React from 'react';
import { useFlashContext } from '../FlashContext';

const FlashExample = () => {
  // use our context hook to get state and the dispatch function
  const [state, dispatch] = useFlashContext();
  const { showFlash, message, style } = state;

  // 3000 milliseconds = 3 seconds
  const delay = ({ time = 3000 }) => {
    return new Promise(resolve => {
      setTimeout(() => {
        resolve(true);
      }, time);
    });
  };

  return (
    <div className="container">
      <div className="row">
        <div className="col">
          <h1>Example Flash Message</h1>
          <p>Press the button to see a flash message!</p>
          <h3>State</h3>
          <p>
            Show Flash: {showFlash.toString()}, Message: {message}, Style:{' '}
            {style}
          </p>
        </div>
      </div>
      <div className="row">
        <div className="col">
          <button
            className="btn btn-primary"
            onClick={async () => {
              dispatch({
                type: 'flash',
                message: 'Hi there!',
                style: 'success',
              });
              await delay(3000);
              dispatch({ type: 'dismiss' });
            }}
          >
            Button
          </button>
        </div>
      </div>
    </div>
  );
};

export default FlashExample;
