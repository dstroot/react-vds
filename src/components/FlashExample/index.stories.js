import React from "react";
import { storiesOf } from "@storybook/react";

import { FlashProvider } from '../FlashContext';
import FlashExample from "../FlashExample";

storiesOf("FlashExample", module)
  .addParameters({
    info: {
      inline: true,
      header: false,
      text: `This supports markdown!
      
      ~~~js
      console.log("hello");
      ~~~
      `
    }
  })
  .add("show FlashExample", () => <FlashProvider>
  <FlashExample />
</FlashProvider>);
