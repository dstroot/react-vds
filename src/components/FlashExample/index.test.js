// After importing the necessary packages, we can write a test which calls
// the render function provided by React Testing Library, which returns an
// object that we can extract asFragment from. Using this we can use the
// built-in functionality from Jest to perform a snapshot test.

import React from 'react';
import { render, screen } from '@testing-library/react';

// component to test
import { FlashProvider } from '../FlashContext';
import FlashExample from '.';

describe('FlashExample', () => {
  it('renders', () => {
    // arrange

    // act
    const { asFragment } = render(
      <FlashProvider>
        <FlashExample />
      </FlashProvider>
    );

    // assert
    expect(asFragment()).not.toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });

  it('it should say Example Flash Message', () => {
    // arrange

    // act
    render(
      <FlashProvider>
        <FlashExample />
      </FlashProvider>
    );
    const h1 = screen.getByText('Example Flash Message');

    // assert
    expect(h1).not.toBeNull();
  });
});
