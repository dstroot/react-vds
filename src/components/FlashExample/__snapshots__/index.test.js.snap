// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlashExample renders 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <h1>
          Example Flash Message
        </h1>
        <p>
          Press the button to see a flash message!
        </p>
        <h3>
          State
        </h3>
        <p>
          Show Flash: false, Message: , Style: 
        </p>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <button
          class="btn btn-success mr-2"
        >
          Success message
        </button>
        <button
          class="btn btn-warning mr-2"
        >
          Warning message 
        </button>
        <button
          class="btn btn-danger"
        >
          Error message
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
